File: 백준/Silver/24954. 물약 구매/물약 구매.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	private static int N;
	private static int[] lst;
	private static int[][][] discountInfo;
	private static boolean[] visited;
	private static int[] selected;
	private static int answer;
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		N = Integer.parseInt(br.readLine());
		
		lst = new int[N+1];
		
		st = new StringTokenizer(br.readLine());
		for (int i = 1; i < N+1; i++) {
			lst[i] = Integer.parseInt(st.nextToken());
		}
		
		discountInfo = new int[N+1][N][2];
		
		for (int i = 1; i < N+1; i++) {
			int pi = Integer.parseInt(br.readLine());
			for (int j = 0; j < pi; j++) {
				st = new StringTokenizer(br.readLine());
				discountInfo[i][j][0] = Integer.parseInt(st.nextToken());
				discountInfo[i][j][1] = Integer.parseInt(st.nextToken());
			}
		} // 입력 완료
		
		visited = new boolean[N+1];
		selected = new int[N];
		
		answer = Integer.MAX_VALUE;
		recur(0, 0, lst);
		
		System.out.println(answer);
	} // end of main

	private static void recur(int cur, int total, int[] arr) {
		if (total > answer) {
			return;
		}
		if (cur == N) {
			answer = Math.min(answer, total);
			return;
		}
		for (int i = 1; i < N+1; i++) {
			if (visited[i]) continue;
			visited[i] = true;
			selected[cur] = i;
			recur(cur + 1, total + arr[i], disco(arr, cur));
			visited[i] = false;
		}
	}

	private static int[] disco(int[] arr, int cur) {
		int[] newArr = new int[N+1];
		for (int i = 1; i < N+1; i++) {
			newArr[i] = arr[i];
		}
		for (int[] dis : discountInfo[selected[cur]]) {
			newArr[dis[0]] -= dis[1];
			if (newArr[dis[0]] <= 0) {
				newArr[dis[0]] = 1;
			}
		}
		return newArr;
	}
} // end of class

File: 백준/Silver/24954. 물약 구매/물약 구매.py
"""
순열 10!이어서 가능

그리고 매번 할인
"""
import sys

input = lambda: sys.stdin.readline().strip()

N = int(input())

lst = [0] + list(map(int, input().split()))

discount = [[]]
for i in range(N):
    c = int(input())
    ls = []
    for j in range(c):
        ls.append(list(map(int, input().split())))
    discount.append(ls)

visited = [False] * (N + 1)
selected = [0] * N
answer = 987654321


def disco(arr, cur):
    new_lst = arr[:]
    result = 0
    result += new_lst[selected[cur]]
    for dis in discount[selected[cur]]:
        new_lst[dis[0]] -= dis[1]
        if new_lst[dis[0]] <= 0:
            new_lst[dis[0]] = 1
    return new_lst


def recur(cur, total, arr):
    global answer
    if total > answer:
        return
    if cur == N:
        # print(selected)
        # new_lst = lst[:]
        # result = 0
        # for a in selected:
        #     result += new_lst[a]
        #     for dis in discount[a]:
        #         new_lst[dis[0]] -= dis[1]
        #         if new_lst[dis[0]] <= 0:
        #             new_lst[dis[0]] = 1
        answer = min(answer, total)
        return
    for i in range(1, N+1):
        if visited[i]:
            continue
        visited[i] = True
        selected[cur] = i
        recur(cur + 1, total + arr[i], disco(arr, cur))
        visited[i] = False


recur(0, 0, lst)
print(answer)